rust/candid/src/de.rs

```diff
@@ -138,29 +132,109 @@ impl<'de> IDLDeserialize<'de> {
         }
         Ok(())
     }
+    /// Return the current DecoderConfig, mainly to extract the remaining quota.
+    pub fn get_config(&self) -> DecoderConfig {
+        self.de.config.clone()
+    }
 }
 
-pub struct Config {
-    zero_sized_values: usize,
+#[derive(Clone)]
+/// Config the deserialization quota, used to prevent spending too much time in decoding malicious payload.
+pub struct DecoderConfig {
+    pub decoding_quota: Option<usize>,
+    pub skipping_quota: Option<usize>,
     full_error_message: bool,
 }
-impl Config {
+impl DecoderConfig {
+    /// Creates a config with no quota. This allows developers to handle large Candid
+    /// data internally, e.g., persisting states to stable memory.
+    /// When using Candid in canister endpoints, we recommend setting the quota to prevent malicious payload.
     pub fn new() -> Self {
         Self {
-            zero_sized_values: 2_000_000,
+            decoding_quota: None,
+            skipping_quota: None,
+            #[cfg(not(target_arch = "wasm32"))]
             full_error_message: true,
+            #[cfg(target_arch = "wasm32")]
+            full_error_message: false,
         }
     }
-    pub fn set_zero_sized_values(&mut self, n: usize) -> &mut Self {
-        self.zero_sized_values = n;
+    /// Limit the total amount of work the deserailizer can perform. Deserialization errors out when the limit is reached.
+    /// If your canister endpoint has variable-length data types and expects that the valid data will be small,
+    /// you can set this limit to prevent spending too much time decoding invalid data.
+    ///
+    /// The cost of decoding a message = 4 * the byte length of the header (the byte before the value part) + the cost of decoding each value.
+    ///
+    /// The cost of decoding a value is roughly defined as follows
+    /// (it's not precise because the cost also depends on how Rust data types are defined),
+    /// ```text
+    /// C : <val> -> <primtype> -> nat
+    /// C(n : nat)      = |leb128(n)|
+    /// C(i : int)      = |sleb128(i)|
+    /// C(n : nat<N>)   = N / 8
+    /// C(i : int<N>)   = N / 8
+    /// C(z : float<N>) = N / 8
+    /// C(b : bool)     = 1
+    /// C(t : text)     = 1 + |t|
+    /// C(_ : null)     = 1
+    /// C(_ : reserved) = 1
+    /// C(_ : empty)    = undefined
+    ///
+    /// C : <val> -> <constype> -> nat
+    /// C(null : opt <datatype>)  = 2
+    /// C(?v   : opt <datatype>)  = 2 + C(v : <datatype>)
+    /// C(?v   : opt <datatype'>) = 2 + C(v : <datatype>) * 50 + 10  // when v cannot be converted to <datatype'>
```

Comment:
```suggestion
    /// C(?v   : opt <datatype'>) = 2 + C(v : <datatype>) * 50 + 10  // when v cannot be converted to <datatype'>
```
where do 50 and 10 come from?

Reply:
These are some arbitrary numbers. 50x is the multiply when we skip values in the decoder. `10` accounts for the cost of restoring states in the backtracking.